python-lib/wordcloud_visualizer.py

```diff
@@ -0,0 +1,267 @@
+        else:
+            # Simply format data similarly
+            df_grouped = [(self.language, df)]
+
+        return df_grouped
+
+    def _tokenize_texts(self, df_grouped: list) -> list:
+        """Private method to tokenize each group of observations in its correct language
+        Args:
+            df_grouped: list of pandas dataframes with one dataframe per language per subchart
+        Returns:
+            List of spacy docs
+        """
+        # Get language and subchart name for each group
+        texts = []
+        group_names = []
+        for name, group in df_grouped:
+            texts.append([group[self.text_column].str.cat(sep=" ")])
+            group_names.append(name)
+
+        # Get tokenization languages differently depending on language/subchart settings combinations
+        if not self.language_column and not self.subchart_column:
+            languages = [self.language]
+        elif self.language_column and self.subchart_column:
+            languages, subcharts = zip(*group_names)
+            languages = list(languages)
+            self.subcharts = list(subcharts)
+        elif self.subchart_column:
+            self.subcharts = group_names
+            languages = [self.language] * len(self.subcharts)
+        else:
+            languages = group_names
+
+        # Tokenize
+        docs = [self.tokenizer.tokenize_list(text, language)[0] for text, language in zip(texts, languages)]
+
+        return docs
+
+    @time_logging(log_message="Counting tokens")
+    def _count_tokens(self, docs: list) -> list:
+        """Private method to count tokens for each document in corpus
+        Args:
+            docs: list of spacy docs on which to count tokens
+        Returns:
+            List of tuples (subchart, counter) where subchart is the subchart the counter belongs to
+        """
+        counters = []
+        for doc in docs:
+            counter = Counter()
+            for token in doc:
+                counter[(token.text)] += 1  # Equivalently, token.lemma_
+            counters.append(counter)
+
+        if not self.subchart_column:
+            # Sum values with same keys
+            counts = Counter()
+            for d in counters:
+                counts.update(d)
+
+            counts = [("", dict(counts))]
+            return counts
+        else:
+            counts = list(zip(self.subcharts, counters))
+            # Aggregate counts by subchart
+            temp_count = {}
+            for subchart, count in counts:
+                if subchart in temp_count.keys():
+                    temp_count[subchart].update(count)
+                else:
+                    temp_count[subchart] = count
+            counts = list(temp_count.items())
+
+            # Remove subcharts emptied by aggregation
+            counts = [(subchart, count) for subchart, count in counts if count != {}]
+            return counts
+
+    def generate_wordclouds(self, counts: list):
```

Comment:
Can you add a more detailed type hint?
```suggestion
    def generate_wordclouds(self, counts: List[Tuple[type1, type2]]) -> Generator[...]:
```